### DISCUSSION TITLE
基础数论（素数筛、gcd、lcm、exgcd、唯一分解定理、快速幂、乘法逆元）

### DISCUSSION THREAD

**Discussion by Lizsen** [CSGrandeur/s-1problem1day1ac#504] [Algorithm]
## 素数

​	定义：若一个正整数无法被除了1和它自身之外的任何自然数整除， 则称该数为素数(或质数)，否则为合数。

​	注意： 在自然数集合中，对于一个足够大的整数 $N$ ， 不超过 $N$ 的质数大约有 $N / InN$ 个。1不是素数。

**素数判定**

通常在算法竞赛用到的判定方法为试除法。

**试除法**

由定义可得，一个数 $x$ 为素数，那么该数有以下性质：

1.    $x >= 2$  
2.   $x$ 不能被 $[2, x - 1]$ 中任何一个数整除

所以我们可以枚举 $[2, x - 1]$ 中每个数 $a$  , 判断是否存在一个数能整除 $x$ ，即`x % a == 0`，若存在，$x$ 为合数，否则为素数。

该算法时间复杂度为*O( $x$ )*，注意到，如果 $[2, x - 1]$ 存在某个数 $a$ 能整除 $x$ ， 那么  $x / a$  也必然也能整除 $x$ ， 所以我们不妨令 $a \leq x / a$ ，整理得到， $a \leq \sqrt{x}$ ， 所以我们只需要枚举 $[2, \sqrt{x}]$ 中每个数即可，时间复杂度为*O($\sqrt{x}$)*。

**代码**

```cpp
bool is_prime(int x) {
	if (x < 2)  return false;
	for (int i = 2; i <= x / i; ++i) {
		if (x % i == 0) return false;
	}
	return true;
}
```

当然，也有更加高效的随机性算法Miller - Robbin，一般情况 $1e9$ 范围内试除法已能胜任， 这里不多介绍。

感兴趣的话，☞[Miller-Robbin素数测试](https://www.cnblogs.com/wolfloraloi/p/11708604.html)

## 素数筛

**素数筛选问题**： 求出 $[2, n]$ 之间所有的素数。

若使用上述的试除法，时间复杂度为*O( $n\sqrt{n}$ )*，时间复杂度不够理想。

下面介绍两种常用的筛法：**埃氏筛法， 线性筛法**

**埃氏筛法(Eratosthenes筛法)**

埃氏筛法的基本思想是， 任意整数 $x$ ，它的倍数 $2x， 3x，...$ 都不是素数。

所以，枚举 $x\in[2，n]$ ，把 $x$ 的倍数： $2x, 3x, \lfloor n/ x\rfloor * x$ 标记为合数。 当我们扫描到一个数 $a$ 时，若它尚未被标记，则说明它不是 $[2, a - 1]$ 中任何一个数的倍数，即 $[2, a - 1]$ 中没有一个数能整除 $a$ ，由定义得，数 $a$ 就是素数。该算法的时间复杂度为*O( $nlogn$ )*。

注意到，某些数被重复筛了多次(标记了很多次)， 能再优化吗？

这里我们先提前引入唯一分解定理(算术基本定理)

**唯一分解定理**：任何一个大于1的整数 $N$ ，都可以唯一分解成有限个素数的乘积， $N = p_{1}^{c_{1}}p_2^{c_{2}}...p_{k}^{c_{k}}$ ，其中 $c_{i}$ 为正整数， $p_{i}$ 为素数， 且满足： $p_{1} < p_{2} < ... < p_{k}$  。

由唯一分解定理可得， 任何一个合数都至少含有一个质因子。所以我们可以只用素数去筛。

**代码**

```cpp
int primes[N], cnt;     // primes[]存储所有素数
bool st[N];         // st[x]存储x是否被筛掉

void get_primes(int n)
{
    for (int i = 2; i <= n; i ++ )
    {
        if (st[i]) continue; // 只用质数去筛，是合数就跳过
        primes[cnt ++ ] = i;
        for (int j = i + i; j <= n; j += i)
            st[j] = true;
    }
}
```

假设 $[2, n - 1]$ 之间的质数为： $p_{1}, p_{2}, .. , p_{m}$ ( $m \approx N / lnN$ )。

**那么上述算法的时间复杂度为 *O* $(n / p_{1} + n / p_{2} + ... + n / p_{m})$ = *O*$(n * (1 / p_{1}+ 1 / p_{2} + ... + 1 / p_{m}))$ $\approx$ *O*$(nloglogn)$**。

**线性筛法**

注意到，即使我们只用素数去筛合数，某些数还是会被筛多次， 如 $6$ 就被 $2$ 和 $3$ 这两个素数各筛了一次。有没有一种办法使得对每个合数，我们只筛一次。有的，由唯一分解定理，对一个合数，都有一个最小的质因子，也就是 $N = p_{1}^{c_{1}}p_2^{c{2}}...p_{k}^{c_{k}}$ 中的 $p_{1}$ 。可以只用每个合数的最小质因子去筛掉它， 时间复杂度为*O( $n$ )*。

**代码**

```cpp
bool st[N];
int primes[N], cnt;
void get_primes(int n) {
	for (int i = 2; i <= n; ++i) {
		if (!st[i]) primes[cnt++] = i;

		// 从小到大枚举每个已经获得的素数，用i * primes[j]的最小质因子priems[j]将i * primes[j]筛掉
		for (int j = 0; primes[j] <= n / i; ++j) {
			/* 
				1.如果 i % primes[j] != 0, 那么primes[j] 一定是i * primes[j]的最小质因子
				即，对于i = p1 ^ c1 * p2 ^ c2 * ... pk ^ ck (p1 < p2 < .. < pk)
				我们要证明primes[j] < p1 < ... < pk。
			证明：
				1)首先证明p1, ..,pk中不包含primes[j];
				因为i % primes[j] != 0, 所以p1, ..,pk中不包含primes[j]。
				
				2)再证明primes[j] < p1;
				使用反证法。
				假设primes[j] > p1，那么必然会在枚举到primes[j]之前p1已经出现，那么就会使得下面if语句条件成立而
				break, 不会再枚举到primes[j]了，矛盾。
				证毕。
			 */
			st[primes[j] * i] = 1;
			/*  
				2.如果此时i % primes[j] == 0, 因为我们是从小大枚举每个素数，所以primes[j]是i的最小质因子。
				那么primes[j]必然是i * primes[j]的最小质因子，为了防止被筛多次，需要break。
			 */
			if (i % primes[j] == 0) break;
		}
	}
}
```

## 最大公约数（greatest common divisor， 简称gcd）

公约数：若自然数 $d$ 即是 $a$ 的约数，又是 $b$ 的约数，则称 $d$ 为 $a$ 和 $b$ 的公约数。

最大公约数： $a$ 和 $b$ 的公约数中最大的一个， 记为 $gcd(a, b)$ 。

最大公约数有如下一些性质：

1. $gcd(a, b)=gcd(b, a)$

2. $gcd(a, 0) =  |a|$

3. $gcd(ka, kb) = k * gcd(a, b)$

下面介绍一种求 $gcd$ 的算法：欧几里得算法。

**欧几里得算法**

​	$\forall a, \quad b\in\mathbb{N}, \quad b \neq 0, \quad gcd(a, \quad b) = gcd(b, \quad a {\kern 2pt} mod {\kern 2pt} b)$

此处略过证明步骤，只给出代码，证明☞[最大公约数](https://oi-wiki.org/math/number-theory/gcd/)。

**代码**

```cpp
typedef long long ll;
ll gcd(ll a, ll b) {
	return b ? gcd(b, a % b) : a;
}
```

时间复杂度为*O( $log(a + b)$ )*。

**多个数的最大公约数**

$gcd(x_{1}, x_{2}, .. x_{n}) = gcd(gcd(x_{1}, x_{2}), x_{3}, ..., x_{n})$

## 最小公倍数 （Least common multiple， 简称lcm）

有如下性质： $lcm(a,b) = a * b / gcd(a, b)$ ，证明☞[最小公倍数](https://oi-wiki.org/math/number-theory/gcd/#最小公倍数)。

**代码**

```cpp
typedef long long ll;
ll lcm(ll a, ll b) {
	return a / gcd(a, b) * b; // 写成如此为了防止溢出
}
```

**多个数的最小公倍数**

$lcm(x_{1}, x_{2}, ..., x_{n}) = lcm(lcm(x_{1}, x_{2}), x_{3}, ..., x_{n})$

## 扩展欧几里得算法（exgcd）

在介绍扩展欧几里得算法之前，需要引入一个定理：裴蜀定理，又称贝祖定理（Bézout's lemma），是一个关于最大公约数的定理。

**裴蜀定理**

$设a,{\kern 5pt}b是不全为0的整数， 则存在整数x,{\kern 5pt}y, 使得ax+by=gcd(a,b)$ 。

证明：

1. 在欧几里得算法的递归出口处： $b = 0, {\kern 5pt}gcd(a, 0) = a$ ， 带入等式得到 $a * x + 0*y = gcd(a, 0) = a$ ，此时 $x = 1, y = 0$ 显然为一组解。

2. 当 $b > 0$ , 由欧几里得算法得， $gcd(a,{\kern 5pt} b) = gcd(b, {\kern 5pt}a {\kern 5pt} mod{\kern 5pt} b) \quad 式1$

   假设存在一对 $x, {\kern 5pt} y$ ,  使得 $b * x + (a {\kern 5pt} mod{\kern 5pt} b)*y = gcd(b,{\kern 5pt} a {\kern 5pt} mod{\kern 5pt} b) \quad 式2$ 

   ​		存在一对 $x^{,}, {\kern 5pt} y^{,}$ , 使得 $a * x^{,} + b*y^{,} = gcd(a,{\kern 5pt} b) \quad 式3$

   对式2化简(以a，b为主元)得到：

   ​		$a * y + b * (x - \lfloor a / b \rfloor * y) = gcd(b,{\kern 5pt} a {\kern 5pt} mod{\kern 5pt} b) \quad 式4$

​		联立式1，式3， 式4，得 $x^{,} = y, {\kern 5pt}y^{,} = (x - \lfloor a / b \rfloor * y)$ 。

**上述内容证明了裴蜀定理，同时也给出了x，y得计算方法。 这种算法被称为扩展欧几里得算法。**

**代码**

```cpp
// 求出a * x + b * y = gcd(a, b)的一组解, 并返回gcd(a, b)
// x，y以引用形式传参
int exgcd(int a, int b, int &x, int &y) {
	if (!b) {
		x = 1, y = 0;
		return a;
	}
    // 求出b * x + a % b * y = gcd(b, a % b)的一组解
	int d = exgcd(b, a % b, x, y);
    
	int z = x; // 先保存一下x的值
    // x`= y, y` = x - a / b * y
    x = y, y = z - a / b * y;
	return d;
}
```

**对于 $ax+by=c$ 是否有解的问题（ $c$ 不一定等于 $gcd(a, b)$ ）**。

这里直接给出结论： $ax+by=c$ 有解，当且仅当  $gcd(a,b){\kern 5pt}|{\kern 5pt}c{\kern 5pt}$ ，即`c % gcd(a, b) == 0`。

如果该等式有解，可以先求出 $ax+by=gcd(a,b)$ 中的一组解 $x_{0}, {\kern 5pt}y_{0}$ ，最后同时乘以 $c / gcd(a, b)$ ，

即 $x = x_{0} * c / gcd(a, b), {\kern 5pt} y = y_{0} * c / gcd(a, b)，x, {\kern 5pt}y$ 为 $ax+by=c$ 的一组解。

令 $d = gcd(a, b)$ ，则

方程 $ax+by=c$ 的通解为： $x = c / d * x_{0} + k * b/d,\quad y = c / d * y_{0} - k * a / d, k \in \mathbb{N}$ 。

## 唯一分解定理（算术基本定理）

定义：任何一个大于1的整数 $N$ ，都可以唯一分解成有限个素数的乘积，**$N = p_{1}^{c_{1}}p_2^{c_{2}}...p_{k}^{c_{k}}$**，其中 $c_{i}$ 为正整数， $p_{i}$ 为素数， 且满足： $p_{1} < p_{2} < ... < p_{k}$ 。

根据定义：可以枚举 $[2, n]$ 所有的素数， 如果某个素数 $p_{i}$ 能够整除 $N$ ，就统计其次数 $c_{i}$ ， 时间复杂度为*O( $N$ )*。

注意到， $[2, n]$ 中能够整除 $N$ 的且其次数 $c_{i}\geq2$ 的某个质数 $p_{i}$ ,  ${\kern 5pt}p_{i} \leq \sqrt{N}$ 。所以我们可以先处理 的 $[2, \sqrt{n}]$ 中的质数 $p_{i}$ ，最后再处理 $c_{i} = 1$ 的部分(如果存在的话)，时间复杂度为*O( $\sqrt{N}$ )*。

**代码**

```cpp
vector<pair<int, int>> divide(int x) {
	vector<pair<int, int>> pfac;
	for (int i = 2; i <= x / i; ++i) {
		if (x % i == 0) {
			int c = 0;
			while (x % i == 0) {
				x /= i;
				++c;
			}
			pfac.push_back(make_pair(i, c));
		}
	}
	if (x > 1) pfac.push_back(make_pair(x, 1));
	return pfac;
}
```

对于唯一分解定理，有一些关于约数的推论。

约数定义：若整数 $d$ 能整除 $n$ ，则称 $d$ 是 $n$ 的约数， $n$ 是 $d$ 的倍数。

**推论**

对 $N = p_{1}^{c_{1}}p_2^{c_{2}}...p_{k}^{c_{k}}$ ，

1. N的某个正约数可以表示成： $d = p_{1}^{a_{1}}p_2^{a_{2}}...p_{k}^{a_{3}}, \quad(0 \leq a_{i} \leq c_{i},i \in[1, k])$ 

2. N的正约数个数为： $(c_{1}+1) * (c_{2}+1) * ... * (c_{k}+1)$

3. N的所有正约数之和为： $(1+p_{1}+p_{1}^2+..+p_{1}^{c_{1}}) * (1+p_{2}+p_{2}^2+..+p_{2}^{c_{2}})*... * (1+p_{k}+p_{k}^2+..+p_{k}^{c_{k}})$ （例如12的所有正约数为1，2，3，4，6，12，那么所有正约数之和为1 + 2 + 3 + 4 + 6 + 12 = 28。）

**代码**

```cpp
int cnt_divisors(int x) {
	int res = 1;
	for (int i = 2; i <= x / i; ++i) {
		if (x % i == 0) {
			int c = 0;
			while (x % i == 0) {
				x /= i;
				++c;
			}
			res *= c + 1;
		}
	}
	if (x > 1) res *= 2;
	return res;
}
ll sum_divisors(int x) {
	ll res = 1;
	for (int i = 2; i <= x / i; ++i) {
		if (x % i == 0) {
			int c = 0;
			while (x % i == 0) {
				++c;
				x /= i;
			}
            // 求解i^0 + i^1 + ... + i^c，这一部分也可以用分治求等比数列前c+1项和之和
			ll t = 0;
			for (int j = 0; j <= c; ++j) {
				t = t * i + 1;
			}
			res *= t;
		}
	}
	if (x > 1) res *= 1 + x;
	return res;
}
// 快速幂
ll qmi(ll a, ll b) {
    ll res = 1;
    for (; b > 0; b >>= 1, a = a * a)
        if (b & 1) res = res * a;
    return res;
}
// 分治求等比数列1, p^1, ..., p^c之和
ll get_sum(int p, int c) {
    if (c == 0) return 1;
    if (c & 1) return (1 + qmi(p, (c + 1) / 2)) * get_sum(p, (c - 1) / 2);
    return (1 + qmi(p, (c + 1) / 2)) * get_sum(p, (c - 1) / 2) + qmi(p, c);
}
```

求一个数的所有正约数

**试除法求约数**

```cpp
vector<int> get_divisors(int x) {
	vector<int> res;
	for (int i = 1; i <= x / i; ++i) {
		if (x % i == 0) {
			res.push_back(i);
			if (x / i != i) res.push_back(x / i);
		}
	}
	return res;
}
```

上述代码时间复杂度为*O*( ${\sqrt{x}}$ )。

有时候也可以用先用线性筛筛出 $[2,{\sqrt{x}}]$ 以内的所有质数，再对 $x$ 进行质因数分解，再利用公式 $d = p_{1}^{a_{1}}p_2^{a_{2}}...p_{k}^{a_{3}}, \quad(0 \leq a_{i} \leq c_{i},i \in[1, k])$ 进行dfs求出所有约数。

## 快速幂

如果你没接触过快速幂， 对于计算 $a ^ n$ 这个问题， 你可能会直接循环求解，或调用`<math.h>`的`pow`函数...

但是如果让你计算 $a ^ n\bmod{p}{\kern 5pt} (1\leq a \leq2e9,{\kern 3pt} 1\leq n \leq2e9,{\kern 3pt}1\leq p \leq2e9,)$ ，如果使用第一种方法会超时，使用库函数的话 $a^n$ 可能会溢出，所以取模也不太方便，快速幂就可以同时解决这两个问题。

1. 根据数学常识，每一个十进制数对应一个二进制数，例如 $(11) _ {10} = (1011) _ {2}$ ，每一个十进制数可以唯一表示为若干个次数不重复的2次幂的和，例如 $11 = 8 + 2 + 1 = 2^{3} + 2^{1} + 2^{0}$ ，那么 $a^{11} = a^{8} * a^{2} * a^{1}$ 。
2. 对于取模运算，由以下性质：
![1](https://user-images.githubusercontent.com/96557990/188182878-da487386-d119-4077-bf28-50d49619196e.png)

利用上述两个性质， 我们可以从后往前枚举二进制数的每一位， 如果当前这一位为1，就a当前的值累积到答案中(`res = res * a % p`)，每次令`a = a * a`( $a^1, a^2, a^4, ...$ 如此倍增)，同时对 $p$ 取模，`a = a * a % p`。

上述算法的时间复杂度为 *O( $logn$ )* ，且两个 $2e9$ 级别的数相乘可以用长整型保存，同时解决了时间复杂度过高和取模不方便两个问题。

**代码**

```cpp
// 计算a^b%p
ll quick_power(ll a, ll b, ll p) {
    ll res = 1 % p;
    for (; b > 0; b >>= 1, a = a * a % p)
        // b & 1可以取出b在二进制表示下的最低位， b >>= 1，可以舍弃最低位，二者结合就可以做到枚举b在二进制表示下的每一位
        if (b & 1) res = res * a % p;
    return res;
}
```

## 乘法逆元

定义：若整数 $b，m$ 互质( $b，m$ 互质 $\Longleftrightarrow$ $gcd(b, m) = 1$ )，并且 $b {\kern 3pt}| {\kern 3pt}a$ ，则存在一个整数 $x$ ，使得 $a / b \equiv a * x\pmod{m}$ ，则称 $x$ 为 $b$ 的模 $m$ 乘法逆元，记为 $b^{-1}\pmod{m}$ 。

将 $x$ 代入得 $a / b \equiv a * b^{-1} \equiv a / b * b * b^{-1}\pmod{m}$ ，所以 $b * b^{-1}\equiv 1\pmod{m}$ 。

因为除法并不满足取模运算的分配律 $(a \div b) \bmod{m} \neq ((a \bmod{m}) \div (b \bmod{m} ))\bmod{m}$ ，而 $a / b \pmod{m} \equiv a * b^{-1} \pmod{m}$ ，即这两者 $\bmod{m}$ 结果相同，可以进行转换。

如果上述内容不能理解也没有关系，你只需要知道一件事：**如遇到 $(a \div b) \bmod{m}$ ，我们可以先让a，b各自对m取模后，再计算 $a * b^{-1}\bmod{m}$** 。

接下来介绍几种求逆元的方法。

**费马小定理**

若 $p$ 为素数，且a,p互质，则有 $a^{p - 1} \equiv 1 \pmod{p}$ 。

将上述式子变形， $a^{p-2} * a \equiv 1 \pmod{p}$ ，此时 $a^{p-2} $即为 $a$ 的模p乘法逆元，可以用快速幂求 $a^{p-2}$ 。

注意该定理的使用范围： $p$ 为素数，且 $a,p$ 互质。

**解线性同余方程 $b * x \equiv 1\pmod{m}$**

如果只有 $b, m$ 互质 ，即 $m$ 不一定是质数 ，就需要解上述线性同余方程来求 $b$ 的模 $m$ 乘法逆元。

在解决这个问题之前，先来看一下什么是**线性同余方程**。

给定整数 $a，b，m$ ，求一个整数 $x$ ，满足 $a*x \equiv b\pmod{m}$ （存在无解情况），因为未知数 $x$ 的次数为1，所以称之为一次同余方程，也称线性同余方程。

接下来讨论该方程解的情况：

对该式子变形， $a * x - m * y = b$ ，不妨写成 $a * x + m * y = b$ ，上面在介绍扩展欧几里得算法时已经给出结论：该方程有解当且仅当 $gcd(a,m){\kern 3pt}|{\kern 3pt}b$ ，可以使用扩展欧几里得算法求得一组解。

回归正题，求解 $b * x \equiv 1\pmod{m}$ 。

对其变形， $b * x + m * y = 1$ ，由 $b，m$ 互质， $gcd(b, m) = 1$ ，可得 $gcd(b,m) {\kern 3pt} | {\kern 3pt}1$ 成立，所以该方程一定有解。

**代码**

```cpp
int exgcd(int a, int b, int&x, int& y) {
	if (b == 0) {
		x = 1, y = 0;
		return a;
	}
	int d = exgcd(b, a % b, x, y);
	int z = x;
	x = y, y = z - a / b * y;
	return d;
}
int inv(int b, int m) {
	int x, y;
	exgcd(b, m, x, y);
	return (x % m + m) % m; // 将范围约束在[0,m-1]以内
}
```

以上介绍的两种方法都是 *O( $logn$ )* 级别的。

**线性求逆元**

如果要求 $[1,n]$ 中每个数模 $p$ 的乘法逆元，使用上述两种算法，时间复杂度为 *O( $nlogn$ )* 。

下面介绍一种线性求逆元的方法。

设 $[1,n]$ 中某个数为 $i$ ，求 $i$ 模 $p$ 的乘法逆元。

首先设 $p = k * i + r(1 < r < i < p)$ ， $k$ 为 $p/i$ 的商， $r$ 是余数。

则 $k * i + r \equiv 0\pmod{p}$ 。

等式两边同时乘上 $i$ 和 $r$ 各自模 $p$ 的逆元 $i^{-1}, r^{-1}$ ，得

​		$k * r^{-1} + i^{-1} \equiv 0 \pmod{p}$

移项，得

​		$i^{-1} \equiv -k*r^{-1} \pmod{{p}}$

​		$i^{-1} \equiv -\lfloor p / i \rfloor*(p \bmod{i})^{-1} \pmod{{p}}$

已知 $1^{-1} \equiv 1 \pmod{p}$ ，于是，就可以根据递推式从前往后求出各项逆元。

**代码**

```cpp
inv[1] = 1;
for (int i = 2; i <= n; ++i) {
    inv[i] = (-1ll * p / i + p) * inv[p % i] % p;
}
```

另外在介绍一种 *O( $n$ )* 求阶乘逆元的方法，这个可以在求组合数用到（ *O( $n$ )* 预处理求出 $i! \bmod {p}$ 和 $i!$ 模 $p$ 的乘法逆元， $i \in [1, n]$）。

如， $C_n^m \bmod{p} = ( (n! \bmod{p}) * (m!^{-1} \bmod{p}) * ((n - m)!^{-1} \bmod{p}) )\bmod{p}$

令 $infac[i]$ 表示 $i!$ 模 $p$ 的乘法逆元。

由如下递推公式， $infac[i] = infac[i + 1] *(i+1)$ 。

可以先求出 $n!$ 模 $p$ 的乘法逆元 $infac[n]$ ，再从后往前递推。

**代码**

```cpp
const int maxn = 100001, mod = 1e9 + 7;
int n, fac[maxn], infac[maxn];
ll qmi(ll a, ll b, ll p) {
    ll res = 1 % p;
    for (; b > 0; b >>= 1, a = a * a % p)
        if (b & 1) res = res * a % p;
    return res;
}
void init() {
    fac[0] = infac[0] = 1;
    for (int i = 1; i < maxn; ++i) fac[i] = 1ll * fac[i - 1] * i % mod;
    infac[maxn - 1] = qmi(fac[maxn - 1], mod - 2, mod);
    for (int i = maxn - 2; i; --i)
        infac[i] = 1ll * infac[i + 1] * (i + 1) % mod;
}
int C(int n, int m) {
    if (n < m) return 0;
    return 1ll * fac[n] * infac[m] % mod * infac[n - m] % mod;
}
```

**题单**

| 序号 | 题号                                                         | 标题                               | 标签                                        | 难度 | 题解                                                         |
| ---- | :----------------------------------------------------------- | ---------------------------------- | ------------------------------------------- | :--- | ------------------------------------------------------------ |
| 1    | [P1579](https://www.luogu.com.cn/problem/list?keyword=1579&page=1) | 哥德巴赫猜想（升级版）             | 素数筛                                      | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P1579.cpp) |
| 2    | [P4752](https://www.luogu.com.cn/problem/P4752)              | Divided Prime                      | 试除法判素数                                | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P4752.cpp) |
| 3    | [P3383](https://www.luogu.com.cn/problem/P3383)              | 【模板】线性筛素数                 | 素数筛                                      | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P3383.cpp) |
| 4    | [POJ2689](http://poj.org/problem?id=2689)                    | Prime Distance                     | 素数筛(大数小区间筛素数)                    | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POJ2689.cpp) |
| 5    | [POJ3126](http://poj.org/problem?id=3126)                    | Prime Path                         | 素数筛+bfs                                  | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POJ3126.cpp) |
| 6    | [AcWing197](https://www.acwing.com/problem/content/199/)     | 阶乘分解                           | 素数筛+唯一分解定理                         | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/AcWing197.cpp) |
| 7    | [P1072](https://www.luogu.com.cn/problem/P1072)              | [NOIP2009 提高组] Hankson 的趣味题 | 素数筛+唯一分解定理+dfs                     | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P1072.cpp) |
| 8    | [POJ2739](http://poj.org/problem?id=2739)                    | Sum of Consecutive Prime Numbers   | 素数筛+前缀和                               | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POJ2739.cpp) |
| 9    | [P1832](https://www.luogu.com.cn/problem/P1832)              | A+B Problem（再升级）              | 素数筛+完全背包求方案数                     | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P1832.cpp) |
| 10   | [AcWing3642](https://www.acwing.com/problem/content/3645/)   | 最大公约数和最小公倍数             | gcd，lcm                                    | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/AcWing3642.cpp) |
| 11   | [HDU1019](https://acm.hdu.edu.cn/showproblem.php?pid=1019)   | Least Common Multiple              | lcm                                         | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/HDU1019.cpp) |
| 12   | [HDU2504](https://acm.hdu.edu.cn/showproblem.php?pid=2504)   | 又见GCD                            | gcd                                         | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/HDU2504.cpp) |
| 13   | [HDU5223](https://acm.hdu.edu.cn/showproblem.php?pid=5223)   | GCD                                | lcm+gcd+思维+构造                           | ⭐⭐⭐  | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/HDU5223.cpp) |
| 14   | [P4549](https://www.luogu.com.cn/problem/P4549)              | 【模板】裴蜀定理                   | 裴蜀定理，gcd                               | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P4549.cpp) |
| 15   | [POJ1061](http://poj.org/problem?id=1061)                    | 青蛙的约会                         | exgcd                                       | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POJ1061.cpp) |
| 16   | [P1082](https://www.luogu.com.cn/problem/P1082)              | [NOIP2012 提高组] 同余方程         | exgcd，乘法逆元                             | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P1082.cpp) |
| 17   | [HDU2669](https://acm.hdu.edu.cn/showproblem.php?pid=2669)   | Romantic                           | exgcd                                       | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/HDU2669.cpp) |
| 18   | [POJ1845](http://poj.org/problem?id=1845)                    | Sumdiv                             | 唯一分解定理+约数之和+分治求等比数列前n项和 | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POJ1845.cpp) |
| 19   | [POJ1995](http://poj.org/problem?id=1995)                    | Raising Modulo Numbers             | 快速幂                                      | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/POj1995.cpp) |
| 20   | [P3197](https://www.luogu.com.cn/problem/P3197)              | [HNOI2008]越狱                     | 快速幂                                      | ⭐⭐   | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P3197.cpp) |
| 21   | [P3811](https://www.luogu.com.cn/problem/P3811)              | 【模板】乘法逆元                   | 乘法逆元                                    | ⭐    | [👍](https://gitee.com/lizsen/my-acm-code/blob/master/基础数论/P3811.cpp) |

**参考资料**

[数论基础 - OI Wiki](https://oi-wiki.org/math/number-theory/basic/)

算法竞赛进阶指南—— 李煜东

[Miller-Robbin素数判定](https://www.cnblogs.com/wolfloraloi/p/11708604.html)

**Comment by Luv-Ray** [comment]
费马小定理那里，是  (mod p) 而不是 (mod m) 吧
